runtime-components/non-repudiation/src/main/scala/com/daml/nonrepudiation/KeyRepository.scala

```diff
@@ -5,24 +5,20 @@ package com.daml.nonrepudiation
 
 import java.security.PublicKey
 
+import com.google.common.io.BaseEncoding
+
 import scala.collection.concurrent.TrieMap
-import scala.util.Try
 
 object KeyRepository {
 
-  type Fingerprint <: String
-
   object Fingerprint {
+    def wrap(bytes: Array[Byte]): Fingerprint = new Fingerprint(bytes)
+    def wrap(key: PublicKey): Fingerprint = wrap(Fingerprints.compute(key))
+  }
 
-    def assertFromString(string: String): Fingerprint =
-      string.asInstanceOf[Fingerprint]
-
-    def fromString(string: String): Try[Fingerprint] =
-      Try(assertFromString(string))
-
-    def compute(key: PublicKey): Fingerprint =
-      assertFromString(Base64Fingerprint.compute(key))
-
+  @SuppressWarnings(Array("org.wartremover.warts.ArrayEquals"))
```

Comment:
Do we trigger this via some autoderived equals instance? Should we overwrite equality to do something sensible?

Reply:
I'll have a look if we can do something about it.

Reply:
Value classes cannot redefine `equals` and I believe `==` is simply synthesized by the compiler. @S11001001 do you have suggestions?

Reply:
Do not use a value class for this kind of thing. [They are only good for syntax extension.](https://discuss.daml.com/t/extends-anyval-is-good-for-extension-methods-and-thats-it-scala-ritual/1548)

You can then manually write an `equals` or, if you change to a `case class`, mix in [`SomeArrayEquals`](https://github.com/digital-asset/daml/blob/81dc6b947374886c0ba41fe27fa0546b09659071/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SomeArrayEquals.scala).

If you really want a newtype, use the abstract type with existential implementation pattern we use for all sorts of things around the codebase like `LedgerString`, `NewBoolean`, `com.daml.concurrent.Future`, or `DiscriminatedSubtype`.  No synthetic `equals` implementations are made in those cases; of course this is not much help to you if you _need_ `equals` to work (but this is completely incompatible with using a newtype at all).

Reply:
I went with the abstract type initially but It accepted a simple `Array[Byte]` in its place. I'll have a second look if I can make it work.

Reply:
Used it, this time raw `Array[Byte]` are rejected. No idea what I got wrong the first time.

Addressed by https://github.com/digital-asset/daml/pull/8672/commits/a099a8f618b0769b742920281ddc27e345524b2e